gobblin-modules/gobblin-azkaban/src/main/java/org/apache/gobblin/service/modules/orchestration/AzkabanClient.java

```diff
@@ -88,19 +96,23 @@ protected AzkabanClient(String username,
                           String url,
                           long sessionExpireInMin,
                           CloseableHttpClient httpClient,
-                          SessionManager sessionManager)
+                          SessionManager sessionManager,
+                          ExecutorService executorService)
       throws AzkabanClientException {
     this.username = username;
     this.password = password;
     this.url = url;
     this.sessionExpireInMin = sessionExpireInMin;
     this.httpClient = httpClient;
     this.sessionManager = sessionManager;
+    this.executorService = executorService;
     this.initializeClient();
     this.initializeSessionManager();
+    this.intializeExecutorService();
     this.retryer = RetryerBuilder.<AzkabanClientStatus>newBuilder()
         .retryIfExceptionOfType(InvalidSessionException.class)
-        .withWaitStrategy(WaitStrategies.exponentialWait(10, TimeUnit.SECONDS))
+        .withAttemptTimeLimiter(AttemptTimeLimiters.fixedTimeLimit(10, TimeUnit.SECONDS, this.executorService))
```

Comment:
Is this really needed? I am a little concerned about creating executor service with a fixed thread pool of 300 threads per client. With multiple Azkaban executor instances, this can lead to a lot of threads.

Reply:
The decision would be based on if you want to share the Azkaban producer or not, if you want to share the producer, then 300 is okay, but if you don't want to share, each job request will have its own AzkabanClient, then you could use SingleThreadPool. It depends on the implementation, that's why I give user option to setExecutorService() by themselves if they want to change the default behavior.